### PR TITLE
gnrc_netreg: mbox and arbitrary callback support

### PR DESCRIPTION
**Pull request by miri64**
Adds dispatch capabilities to `gnrc_netapi`/`gnrc_netreg` other than target threads. This is optional behavior. The new target types introduced are:

* `mbox` (see #4919)
* arbitrary callbacks providing both `TYPE` and `pkt`

Depends on ~~#5524~~ (merged), ~~#5525~~ (merged), and ~~#4919~~ (merged)

### CODE REVIEW

**Comment by miri64** [comment]
The callback idea was mainly to have some experiments done (and maybe future option provided) to run GNRC in a single thread.

**Comment by miri64** [comment]
And it can also be used to implement receive callbacks in `conn` very slim (see #5091)

**Comment by miri64** [comment]
Rebased to current master.

**Comment by miri64** [comment]
Fixed some errors.

**Comment by miri64** [comment]
Rebased to current dependencies

**Comment by miri64** [comment]
Rebased to current #5524 (as it is the only remaining dependency) and squashed.

**Comment by miri64** [comment]
Rebased to current #5524 and fixed a missing change.

**Comment by kaspar030** [comment]
Quite a lot of changes, and I don't know gnrc well enough. Will reassign.

**Comment by miri64** [comment]
@kaspar030 that's why I divided it up into two PRs ;-) (the other one being integrated here and being #5524)

**Comment by miri64** [comment]
@haukepetersen `gnrc_netreg` was basically your brain child, too ;). Care for a review if @OlegHahm does not find the time?

**Comment by miri64** [comment]
Oh, #5524 needs merging first ^^"

**Comment by miri64** [comment]
Needed for #5772, so added to release.

**Comment by miri64** [comment]
Rebased to current #5526

File: Makefile.dep [hunk 1]
```diff
@@ -59,6 +59,10 @@ ifneq (,$(filter gnrc_conn_udp,$(USEMODULE)))
   USEMODULE += gnrc_udp
 endif
 
+ifneq (,$(filter gnrc_netreg_extra,$(USEMODULE)))
```
Comment (miri64) [inline Makefile.dep]: Well the `_extra` module pulls in the mbox and callback functionality.

Comment (miri64) [inline Makefile.dep]: I can fix it up to devide it into two modules, but the complexity is growing in both instances.

**Comment by haukepetersen** [comment]
looks valid to me, the only thing is the name of the `MODULE_GNRC_NETREG_EXTRA` module. I would prefer `MODULE_GNRC_NETAPI_EXTRA`, as this is the more general term and in the end this PR provides us with more flexibility when using netapi.

File: Makefile.dep (hunk 1 above)
Comment (miri64) [inline Makefile.dep]: Introduced two pseudomodule with a better name.

**Comment by haukepetersen** [comment]
Oh, and of course once squashed and rebased...

**Comment by miri64** [comment]
will wait for #5524 to get merged with the squash (+ rebase), so I can cleanly rebase #5772 on this one.

**Comment by miri64** [comment]
Rebased and squashed

**Comment by miri64** [comment]
Fixed issue in `tests/slip` and squashed immediately.

**Comment by cgundogan** [comment]
Murdock is happy => GO

**Pull request merged**